src/macros-by-example.adoc

```diff
@@ -0,0 +1,827 @@
+therein is our first non-trivial _template_, an expression in Ion’s new domain-specific language
+for defining macro functions.  This expression language treats Ion scalar values (except for
+symbols) as literals, giving the decimal in ``pi``’s template its intended meaning.  Expressions
+that are structs are interpreted as quasi-literals: the field names are literal, but the field
+“values” are arbitrary expressions.  This is why the `amount` and `currency` field names show up
+as-is in the expansion.  The sub-expressions in this template demonstrate that the expression
+language treats symbols as variable references.  Here, `a` and `c` in the template refer to the
+parameters of the macro, and during expansion they are “filled in” with the values supplied by
+the invoking E-expression.
+
+The template language also treats lists quasi-literally, and every element inside the list is an
+expression.  Here’s a silly macro to illustrate:
+
+[{mrk}]
+----
+(*public* reverse [a, b] [b, a])
+----
+----
+(:reverse first {amount:a, currency:c}) ⇒ [{amount:a, currency:c}, first]
+----
+
+To point out what may be obvious, note that the `a` in the E-expression is _not_ part of the
+expression language and not a reference to the macro’s first parameter or any other named entity.
+From the point of view of ``reverse``’s template, the input struct is literal data.
+
+E-expressions can nest, so we could also encode the same data using `price`:
+
+----
+(:reverse first (:price a c)) ⇒ [{amount:a, currency:c}, first]
+----
+
+
+=== Invoking Macros from Templates
+
+So the template language treats symbols as parameter references, other scalars as literals, and
+structs and lists as quasi-literals, but what about S-expressions?
+
+[{mrk}]
+----
+(*public* website_url
+  [path]
+  (make_string "https://www.amazon.com/" path))
+----
+
+Template expressions that are S-expressions are operator invocations, where the operators are
+either special forms or macro invocations.  In this case, it’s the latter: an invocation of the
+system macro (_i.e._, built-in function) `make_string`, which concatenates its arguments to produce
+a single string:
+
+----
+(:website_url "gp/cart") ⇒ "https://www.amazon.com/gp/cart"
+----
+
+In the template language, macro invocations can appear almost anywhere:
+
+[{mrk}]
+----
+(*public* detail_page_url
+  [asin]
+  (website_url (make_string "dp/" asin)))
+----
+----
+(:detail_page_url "B08KTZ8249") ⇒ "https://www.amazon.com/dp/B08KTZ8249"
+----
+
+NOTE: While this doesn’t look like much of an improvement, the full string takes 38 bytes to encode,
+but the macro invocation takes as few as 12 bytes.
+
+
+=== Special Form: `*literal*`
+
+To explain more carefully, when a template-expression is syntactically an S-expression, its first
+element must be a symbol that matches either a set of keywords denoting the special forms, or the
+name of a previously-defined macro.  The interpretation of the S-expression’s remaining elements
```

Comment:
> or the name of a previously defined macro

Is this something different than the `(:` syntax to invoke a macro? (Or in other words, are we conflating S-expressions and E-expressions here?)

Reply:
I'm adding a paragraph to the very top talking about the DSL-as-data aspect, and a new subsection here talking more about the distinctions.